### ISSUE TITLE
[Question] conan with development workflows

### ISSUE THREAD

**Issue by danielhalati** [conan-io/conan#5539]
Hello,

I want to integrate conan at my company and I have some questions. I will detail a little bit our development workflows and ask some questions to find out how to fit conan here.

Our final software product is a desktop application which has a lot of dependencies of inhouse developed libraries and 3rd party libraries: depends on libA, libB, libC, Qt, boost and so on.

conan will show up here and we will have all these dependencies listed inside conan.txt. All the dependencies that are developed by us will be also handled by conan.

Assuming we need to provide a fix inside libA we create a branch (this will create automatically a Jenkins build) and then open a pull request to merge the fix inside the release branch.

After fixing a bug we have the necessity to provide the whole software compile with the fix so the testing department can take over and approve the fix (we cannot merge the pull request until they approve). In order to provide the correct software package to the testing department, I need to go inside every repository which has a dependency of libA and edit its conanfile.txt so it retrieves another package with the fix inside.

To me, this approach can get cumbersome since various components can depend on the same library. It will be very confusing for a developer to know who depends on the library he provided a fix for so he can go inside that repository and create a branch which alters conanfile.txt

On another topic: can the conan channel correspond to a GIT branch?

**Comment by danimtb** [comment]
Hi @danielhalati 

Yes, the scenario you describe is a very common development flow.

Normally you will create a new branch or PR in order to make the changes to fix the issue in LibA. Those changes would normally fire different CI jobs in order to build packages with the fix. The usual recommendation is to upload those packages to a new Artifactory repository with the same name of the branch or PR. That way the testing/QA team can get the packages from that isolated repository.

Changing the channel with the PR/branch name could be another possibility, but as you pointed our
it would be difficult to bump the references on all the recipes.

I think the best way would be the idea of the Artifactory repositories per branch and having a package ID mode with revisions to know exactly what to rebuild downstream in CI https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#versioning-schema

**Comment by Johnnyxy** [comment]
@danielhalati 
If you have to build a pre-release version for your testing teams with at least changes to any build chain files then like already recommended it would be best to do it in a separate repository and use the same names/channels. This way you can then let the fix be approved and then merged into the "real" repository.

For the other approch with Git-branch-names as Conan channels:
Please be aware that there are naming rules you have to follow.
For one there are Conan's naming rules, see here: https://github.com/conan-io/conan/blob/9f3e486f98c72c1aaec16c0692ec711d0bae2cd6/conans/model/ref.py#L90
But there can be other systems like Jenkins which sometimes have problems, like using branch names in paths, etc. (e.g. see [here](https://issues.jenkins-ci.org/browse/JENKINS-34564), [here](https://stackoverflow.com/a/39616225)).

**Comment by danielhalati** [comment]
> Hi @danielhalati
> 
> Yes, the scenario you describe is a very common development flow.
> 
> Normally you will create a new branch or PR in order to make the changes to fix the issue in LibA. Those changes would normally fire different CI jobs in order to build packages with the fix. The usual recommendation is to upload those packages to a new Artifactory repository with the same name of the branch or PR. That way the testing/QA team can get the packages from that isolated repository.
> 
> Changing the channel with the PR/branch name could be another possibility, but as you pointed our
> it would be difficult to bump the references on all the recipes.
> 
> I think the best way would be the idea of the Artifactory repositories per branch and having a package ID mode with revisions to know exactly what to rebuild downstream in CI https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#versioning-schema

Thanks for the reply. Your approach sounds good, but I think it doesn't help me because I only use conan to package existing binaries. Our inhouse toolchain is responsible to build the binaries.

**Comment by danimtb** [comment]
@danielhalati you can still build the packages with your toolchain and upload them to the Artifactory repository matching your git branch. It is totally possible.

However, without building the packages directly from the Conan recipe, you lose the possibility of dynamically building only the downstream affected packages in the graph with build policies like ``--build missing`` or ``--build cascade``, but that is another and more advanced issue.

**Comment by danimtb** [comment]
I am closing this question as solved as there was no activity but please feel free to reopen if you need more assistance or open a new one. Thanks!
